src/components/Dropdown/Dropdown.tsx

```diff
@@ -0,0 +1,217 @@
+}: Props) {
+  const wrapperRef = createRef<HTMLDivElement>()
+  const dropdownRef = createRef<HTMLOListElement>()
+  const [items, setItems] = useState(selectedItems)
+  const [searchTerm, setSearchTerm] = useState('')
+  const debouncedSearchTerm = useDebounce(searchTerm, 200)
+  const [matchingItems, setMatchingItems] = useState<Item[]>([])
+  const [doShowResultsList, setDoShowResultsList] = useState(false)
+  const [suggestionIndex, setSuggestionIndex] = useState(-1)
+  useOutsideClickAlerter(wrapperRef, () => setDoShowResultsList(false))
+
+  useEffect(() => {
+    const onSearchTermChange = async (searchTerm: string) => {
+      const matchingItems = await getItems(searchTerm)
+      setMatchingItems(matchingItems)
+      if (searchTerm.length > 1) {
+        setDoShowResultsList(true)
+      }
+    }
+
+    onSearchTermChange(debouncedSearchTerm)
+  }, [debouncedSearchTerm, getItems])
+
+  useEffect(() => {
+    setSelectedItems(items)
+  }, [items, setSelectedItems])
+
+  const setSingleItem = (item: Item) => {
+    for (const i of matchingItems) {
+      if (i.id !== item.id) {
+        deselectItem(i)
+      }
+    }
+    selectItem(item)
+  }
+
+  const toggleItem = (item: Item) => {
+    const isSelected = selectedItems.some(
+      (selectedItem: Item) => selectedItem.id === item.id
+    )
+
+    if (isSingleSelectable) {
+      setSingleItem(item)
+    } else if (isSelected) {
+      deselectItem(item)
+    } else {
+      selectItem(item)
+    }
+  }
+
+  const handleOnChange = (event: React.FormEvent<HTMLInputElement>) => {
+    setSuggestionIndex(-1)
+    setSearchTerm(event.currentTarget.value)
+  }
+
+  const selectItem = (selectedItem: Item) => {
+    setItems((previousState) => [...new Set([...previousState, selectedItem])])
+  }
+
+  const deselectItem = (selectedItem: Item) =>
+    setItems((previousState) =>
+      previousState.filter((item) => item.id !== selectedItem.id)
+    )
+
+  const resultsToShow = Math.min(matchingItems.length, maxAmountOfItems)
+
+  const handleKeyDown: KeyboardEventHandler<HTMLInputElement> = (e) => {
+    if (e.key === 'Escape' || e.key === 'Esc' || e.key === 'Tab') {
+      setDoShowResultsList(false)
+      setSuggestionIndex(-1)
+    }
+    if (e.key === 'Enter') {
+      e.preventDefault()
+      const item = matchingItems[suggestionIndex]
+      if (item) {
+        toggleItem(item)
+      }
+    }
+    if (e.key === 'ArrowDown' || e.key === 'Down') {
+      e.preventDefault()
+      setDoShowResultsList(true)
+      if (suggestionIndex === -1) {
+        setSuggestionIndex(0)
+        return
+      }
+
+      const isGreaterThanResultsListLength = suggestionIndex < resultsToShow - 1
+      setSuggestionIndex(
+        isGreaterThanResultsListLength ? suggestionIndex + 1 : 0
+      )
+    }
+    if (e.key === 'ArrowUp' || e.key === 'Up') {
+      e.preventDefault()
+      setDoShowResultsList(true)
+      if (suggestionIndex === -1) {
+        setSuggestionIndex(0)
```

Comment:
Fair enough. Is this some kind of recommendation to go "down" to the first item in the list by pressing 'up'?

Reply:
Yes, I think it's common practice for an input field to just open the options, and not also go down the list.